src/core/thread/network_data_publisher.cpp

```diff
@@ -0,0 +1,1079 @@
+{
+    // Removes the service entry from network data (if it was added).
+
+    bool registerWithLeader = false;
+
+    VerifyOrExit((GetState() == kAdded) || (GetState() == kRemoving));
+
+    switch (GetType())
+    {
+    case kTypeAnycast:
+        SuccessOrExit(Get<Service::Manager>().Remove<Service::DnsSrpAnycast>(
+            Service::DnsSrpAnycast::ServiceData(mInfo.GetSequenceNumber())));
+        break;
+
+    case kTypeUnicast:
+        SuccessOrExit(Get<Service::Manager>().Remove<Service::DnsSrpUnicast>(
+            Service::DnsSrpUnicast::ServiceData(mInfo.GetAddress(), mInfo.GetPort())));
+        break;
+
+    case kTypeUincastMeshLocalEid:
+        SuccessOrExit(Get<Service::Manager>().Remove<Service::DnsSrpUnicast>());
+        break;
+    }
+
+    registerWithLeader = true;
+    Notify(kEventEntryRemoved);
+
+exit:
+    SetState(aNextState);
+    return registerWithLeader;
+}
+
+void Publisher::DnsSrpServiceEntry::Notify(Event aEvent) const
+{
+#if OPENTHREAD_CONFIG_SRP_SERVER_ENABLE
+    Get<Srp::Server>().HandleNetDataPublisherEvent(aEvent);
+#endif
+
+    if (mCallback != nullptr)
+    {
+        mCallback(static_cast<otNetDataPublisherEvent>(aEvent), mCallbackContext);
+    }
+}
+
+void Publisher::DnsSrpServiceEntry::Process(void)
+{
+    // This method checks the entries currently present in Network Data
+    // based on which it then decides whether or not take action
+    // (add/remove or keep monitoring).
+
+    uint8_t numEntries          = 0;
+    uint8_t numPreferredEntries = 0;
+    uint8_t desiredNumEntries   = 0;
+
+    // Do not make any changes if device is not attached, and wait
+    // for role change event.
+    VerifyOrExit(Get<Mle::Mle>().IsAttached());
+
+    VerifyOrExit(GetState() != kNoEntry);
+
+    switch (GetType())
+    {
+    case kTypeAnycast:
+        CountAnycastEntries(numEntries, numPreferredEntries);
+        desiredNumEntries = kDesiredNumAnycast;
+        break;
+
+    case kTypeUnicast:
+    case kTypeUincastMeshLocalEid:
+        CountUnicastEntries(numEntries, numPreferredEntries);
+        desiredNumEntries = kDesiredNumUnicast;
+        break;
+    }
+
+    UpdateState(numEntries, numPreferredEntries, desiredNumEntries);
+
+exit:
+    return;
+}
+
+void Publisher::DnsSrpServiceEntry::CountAnycastEntries(uint8_t &aNumEntries, uint8_t &aNumPreferredEntries) const
+{
+    // Count the number of matching "DNS/SRP Anycast" service entries
+    // in the Network Data (the match requires the entry to use same
+    // "sequence number" value). We prefer the entries associated with
+    // smaller RLCO16.
+
+    Service::DnsSrpAnycast::ServiceData serviceData(mInfo.GetSequenceNumber());
+    const ServiceTlv *                  serviceTlv = nullptr;
+
+    while ((serviceTlv = Get<Leader>().FindNextMatchingService(serviceTlv, Service::kThreadEnterpriseNumber,
+                                                               reinterpret_cast<const uint8_t *>(&serviceData),
```

Comment:
Here we internationally want to to include the "seq number" when matching the service data (i.e., we want both "service number" and "seq number" to be present and match). If we see an existing DNS/SRP anycast service entry in network data with a different "seq number", then it is not a match with our entry and we should not count it as such (it may be an older seq number).

Regarding forward compatibility, this is already handled since we are using `FindNextMatchingService()` which does a partial match of service data (it accepts the service entry if service data is 2 or more).

Reply:
IIUC, `sequence number` aims to trigger re-registration in some situations, it may change over time for some reason,  we may always prefer larger `sequence number`.

Reply:
That's correct. That's how the current SRP client code uses the entries from network data (the one with higher seq number is used). How the seq number is determined would be up to SRP server (e.g. to discover and sync with other servers  - which I guess would be part of the replication mechanism (to be determined)).

However, I think the `Publisher` should not care or be aware of any of this (the same way that it is not determining/managing the prefix (e.g., OMR prefix) and/or external route entries). The `Publisher` is instructed to publish certain entries and its responsibility is to determine when/if to add/remove the published entry in network data (by monitoring it for matching entries)